src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs

```diff
@@ -77,46 +77,83 @@ private static async Task IndexingAsync(ShellScope scope, IEnumerable<ContentCon
+                                await indexManager.DeleteDocumentsAsync(indexSettings.IndexName, new string[] { context.ContentItem.ContentItemId });
+                            }
                         }
                         else
                         {
-                            var buildIndexContext = new BuildIndexContext(new DocumentIndex(contentItem.ContentItemId, contentItem.ContentItemVersionId), contentItem, new string[] { contentItem.ContentType }, new LuceneContentIndexSettings());
-                            await contentItemIndexHandlers.InvokeAsync(x => x.BuildIndexAsync(buildIndexContext), logger);
-
-                            await luceneIndexManager.DeleteDocumentsAsync(indexSettings.IndexName, new string[] { contentItem.ContentItemId });
-                            await luceneIndexManager.StoreDocumentsAsync(indexSettings.IndexName, new DocumentIndex[] { buildIndexContext.DocumentIndex });
+                            if (context is PublishContentContext publishContext)
+                            {
+                                if (publishContext.Cancel)
+                                {
+                                    continue;
+                                }
+
+                                if (publishContext.IsPublishing == true)
+                                {
+                                    await StoreDocument(publishContext.PublishingItem, contentItemIndexHandlers,indexManager, logger, indexSettings);
+                                }
+                                else
+                                {
+                                    await indexManager.DeleteDocumentsAsync(indexSettings.IndexName, new string[] { context.ContentItem.ContentItemId });
+                                }
+                            }
+                            else if (context is UpdateContentContext updateContext)
+                            {
+                                await StoreDocument(updateContext.UpdatingItem, contentItemIndexHandlers, indexManager, logger, indexSettings);
+                            }
+                            else if (context is CreateContentContext createContext)
+                            {
+                                await StoreDocument(createContext.CreatingItem, contentItemIndexHandlers, indexManager, logger, indexSettings);
+                            }
                         }
                     }
                 }
             }
         }
+
+        private static async Task StoreDocument(ContentItem contentItem, IEnumerable<IContentItemIndexHandler> contentItemIndexHandlers, LuceneIndexManager luceneIndexManager, ILogger logger, LuceneIndexSettings indexSettings)
+        {
+
+            var buildIndexContext = new BuildIndexContext(new DocumentIndex(contentItem.ContentItemId, contentItem.ContentItemVersionId), contentItem, new string[] { contentItem.ContentType }, new LuceneContentIndexSettings());
```

Comment:
Create a local `new string[] { contentItem.ContentType }` and reuse the same instance in both method calls

Reply:
The other one is `new string[] { contentItem.ContentItemId }`